### PR TITLE
[3.11] Trim trailing whitespace and test on CI (GH-104275)

### PR DESCRIPTION
**Pull request by hugovk**

### CODE REVIEW

**Comment by hugovk** [comment]
I'll follow this with a combined backport of https://github.com/python/cpython/pull/105750 and https://github.com/python/cpython/pull/108212.

I left them out of this one to keep things separate and clearer.

File: Lib/email/__init__.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
     ]
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/__init__.py]: PEP8 (we don't need 3 blank lines here):

```suggestion
```

File: Lib/email/base64mime.py [hunk 2]
```diff
@@ -45,7 +45,7 @@
 MISC_LEN = 7
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/base64mime.py]:
```suggestion
```

File: Lib/email/base64mime.py [hunk 3]
```diff
@@ -57,7 +57,7 @@ def header_length(bytearray):
     return n
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/base64mime.py]:
```suggestion
```

File: Lib/email/base64mime.py [hunk 4]
```diff
@@ -72,7 +72,7 @@ def header_encode(header_bytes, charset='iso-8859-1'):
     return '=?%s?b?%s?=' % (charset, encoded)
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/base64mime.py]:
```suggestion
```

File: Lib/email/base64mime.py [hunk 5]
```diff
@@ -98,7 +98,7 @@ def body_encode(s, maxlinelen=76, eol=NL):
     return EMPTYSTRING.join(encvec)
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/base64mime.py]:
```suggestion
```

File: Lib/email/charset.py [hunk 6]
```diff
@@ -18,7 +18,7 @@
 from email.encoders import encode_7or8bit
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/charset.py]:
```suggestion
```

File: Lib/email/charset.py [hunk 7]
```diff
@@ -32,7 +32,7 @@
 EMPTYSTRING = ''
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/charset.py]:
```suggestion
```

File: Lib/email/charset.py [hunk 8]
```diff
@@ -104,7 +104,7 @@
     }
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/charset.py]:
```suggestion
```

File: Lib/email/charset.py [hunk 9]
```diff
@@ -153,7 +153,7 @@ def add_codec(charset, codecname):
     CODEC_MAP[charset] = codecname
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/charset.py]:
```suggestion
```

File: Lib/email/charset.py [hunk 10]
```diff
@@ -163,7 +163,7 @@ def _encode(string, codec):
         return string.encode(codec)
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/charset.py]:
```suggestion
```

File: Lib/email/encoders.py [hunk 11]
```diff
@@ -16,7 +16,7 @@
 from quopri import encodestring as _encodestring
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/encoders.py]:
```suggestion
```

File: Lib/email/encoders.py [hunk 12]
```diff
@@ -34,7 +34,7 @@ def encode_base64(msg):
     msg['Content-Transfer-Encoding'] = 'base64'
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/encoders.py]:
```suggestion
```

File: Lib/email/encoders.py [hunk 13]
```diff
@@ -46,7 +46,7 @@ def encode_quopri(msg):
     msg['Content-Transfer-Encoding'] = 'quoted-printable'
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/encoders.py]:
```suggestion
```

File: Lib/email/encoders.py [hunk 14]
```diff
@@ -64,6 +64,6 @@ def encode_7or8bit(msg):
         msg['Content-Transfer-Encoding'] = '7bit'
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/encoders.py]:
```suggestion
```

File: Lib/email/feedparser.py [hunk 15]
```diff
@@ -41,7 +41,7 @@
 NeedMoreData = object()
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/feedparser.py]:
```suggestion
```

File: Lib/email/feedparser.py [hunk 16]
```diff
@@ -132,7 +132,7 @@ def __next__(self):
         return line
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/feedparser.py]:
```suggestion
```

File: Lib/email/generator.py [hunk 17]
```diff
@@ -22,7 +22,7 @@
 fcre = re.compile(r'^From ', re.MULTILINE)
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/generator.py]:
```suggestion
```

File: Lib/email/generator.py [hunk 18]
```diff
@@ -392,7 +392,7 @@ def _make_boundary(cls, text=None):
     def _compile_re(cls, s, flags):
         return re.compile(s, flags)
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/generator.py]: (This blank line should stay, or there'll only be one blank line in between the class definition and the function above it)

File: Lib/email/generator.py [hunk 19]
```diff
@@ -443,7 +443,7 @@ def _compile_re(cls, s, flags):
         return re.compile(s.encode('ascii'), flags)
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/generator.py]:
```suggestion
```

File: Lib/email/generator.py [hunk 20]
```diff
@@ -503,7 +503,7 @@ def _dispatch(self, msg):
                     }, file=self)
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/generator.py]:
```suggestion
```

File: Lib/email/header.py [hunk 21]
```diff
@@ -52,12 +52,12 @@
 _embedded_header = re.compile(r'\n[^ \t]+:')
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/header.py]:
```suggestion
```

File: Lib/email/header.py [hunk 22]
```diff
@@ -52,12 +52,12 @@
...
-

+
 # Helpers
 _max_append = email.quoprimime._max_append
 
 
-

+
```
Comment (AlexWaygood) [inline Lib/email/header.py]:
```suggestion
```

File: Lib/email/header.py [hunk 23]
```diff
@@ -152,7 +152,7 @@ def decode_header(header):
     return collapsed
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/header.py]:
```suggestion
```

File: Lib/email/header.py [hunk 24]
```diff
@@ -175,7 +175,7 @@ def make_header(decoded_seq, maxlinelen=None, header_name=None,
     return h
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/header.py]:
```suggestion
```

File: Lib/email/header.py [hunk 25]
```diff
@@ -409,7 +409,7 @@ def _normalize(self):
         self._chunks = chunks
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/header.py]:
```suggestion
```

File: Lib/email/iterators.py [hunk 26]
```diff
@@ -15,7 +15,7 @@
 from io import StringIO
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/iterators.py]:
```suggestion
```

File: Lib/email/iterators.py [hunk 27]
```diff
@@ -29,7 +29,7 @@ def walk(self):
             yield from subpart.walk()
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/iterators.py]:
```suggestion
```

File: Lib/email/iterators.py [hunk 28]
```diff
@@ -55,7 +55,7 @@ def typed_subpart_iterator(msg, maintype='text', subtype=None):
                 yield subpart
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/iterators.py]:
```suggestion
```

File: Lib/email/mime/base.py [hunk 29]
```diff
@@ -11,7 +11,7 @@
 from email import message
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/mime/base.py]:
```suggestion
```

File: Lib/email/mime/message.py [hunk 30]
```diff
@@ -10,7 +10,7 @@
 from email.mime.nonmultipart import MIMENonMultipart
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/mime/message.py]:
```suggestion
```

File: Lib/email/mime/multipart.py [hunk 31]
```diff
@@ -9,7 +9,7 @@
 from email.mime.base import MIMEBase
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/mime/multipart.py]:
```suggestion
```

File: Lib/email/mime/nonmultipart.py [hunk 32]
```diff
@@ -10,7 +10,7 @@
 from email.mime.base import MIMEBase
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/mime/nonmultipart.py]:
```suggestion
```

File: Lib/email/mime/text.py [hunk 33]
```diff
@@ -10,7 +10,7 @@
 from email.mime.nonmultipart import MIMENonMultipart
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/mime/text.py]:
```suggestion
```

File: Lib/email/parser.py [hunk 34]
```diff
@@ -67,15 +67,15 @@ def parsestr(self, text, headersonly=False):
         return self.parse(StringIO(text), headersonly=headersonly)
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/email/parser.py]:
```suggestion
```

File: Lib/email/parser.py [hunk 35]
```diff
@@ -67,15 +67,15 @@ def parsestr(self, text, headersonly=False):
...
 class HeaderParser(Parser):
     def parse(self, fp, headersonly=True):
         return Parser.parse(self, fp, True)
 
     def parsestr(self, text, headersonly=True):
         return Parser.parsestr(self, text, True)
 
-

+
```
Comment (AlexWaygood) [inline Lib/email/parser.py]: (Keep this blank line)

File: Lib/test/test_isinstance.py [hunk 36]
```diff
@@ -8,7 +8,7 @@
 from test import support
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/test/test_isinstance.py]:
```suggestion
```

File: Lib/test/test_isinstance.py [hunk 37]
```diff
@@ -147,7 +147,7 @@ def getbases(self):
         self.assertRaises(TypeError, issubclass, B, C())
 
 
-
+
```
Comment (AlexWaygood) [inline Lib/test/test_isinstance.py]:
```suggestion
```

File: Modules/_io/bufferedio.c [hunk 38]
```diff
@@ -1740,7 +1740,7 @@ _bufferedreader_peek_unlocked(buffered *self)
     self->pos = 0;
     return PyBytes_FromStringAndSize(self->buffer, r);
 }
-
+
```
Comment (AlexWaygood) [inline Modules/_io/bufferedio.c]:
```suggestion
```

File: Modules/_io/bufferedio.c [hunk 39]
```diff
@@ -2047,7 +2047,7 @@ _io_BufferedWriter_write_impl(buffered *self, Py_buffer *buffer)
     LEAVE_BUFFERED(self)
     return res;
 }
-
+
```
Comment (AlexWaygood) [inline Modules/_io/bufferedio.c]:
```suggestion
```

File: Modules/_io/bufferedio.c [hunk 40]
```diff
@@ -2253,7 +2253,7 @@ bufferedrwpair_closed_get(rwpair *self, void *context)
     }
     return PyObject_GetAttr((PyObject *) self->writer, &_Py_ID(closed));
 }
-
+
```
Comment (AlexWaygood) [inline Modules/_io/bufferedio.c]:
```suggestion
```

File: Tools/i18n/pygettext.py [hunk 41]
```diff
@@ -174,7 +174,7 @@
 EMPTYSTRING = ''
 
 
-
+
```
Comment (AlexWaygood) [inline Tools/i18n/pygettext.py]:
```suggestion
```

File: Tools/i18n/pygettext.py [hunk 42]
```diff
@@ -196,15 +196,15 @@
...
 def usage(code, msg=''):
     print(__doc__ % globals(), file=sys.stderr)
     if msg:
         print(msg, file=sys.stderr)
     sys.exit(code)
 
 
-

+
```
Comment (AlexWaygood) [inline Tools/i18n/pygettext.py]:
```suggestion
```

File: Tools/i18n/pygettext.py [hunk 43]
```diff
@@ -515,7 +515,7 @@ def write(self, fp):
                 print('msgstr ""\n', file=fp)
 
 
-
+
```
Comment (AlexWaygood) [inline Tools/i18n/pygettext.py]:
```suggestion
```

**Review by AlexWaygood — COMMENTED** [review]
A lot of this diff is removing invisible pagebreak characters. For most of these cases, we should delete the whole line rather than just stripping the "trailing whitespace" from the line:

**Comment by AlexWaygood** [comment]
See #102531 for where we removed invisible pagebreak characters from the `main` branch. (@hugovk, is it worth seeing if that PR could just be cleanly backported to 3.11?)

**Comment by hugovk** [comment]
Good idea, and it backported cleanly :tada: https://github.com/python/cpython/pull/108266

**Comment by AlexWaygood** [comment]
> Good idea, and it backported cleanly 🎉 #108266

And merged! You now have a bit of a merge conflict ;)

**Comment by hugovk** [comment]
Reverted [Trim trailing whitespace](https://github.com/python/cpython/pull/108215/commits/cd547a2fd51455f9816bfba8e6912a2a4ca8155f), will update 3.11 into here, and then see what's left to fix.

**Comment by hugovk** [comment]
And autofixed via pre-commit, let's check these changes.

**Comment by hugovk** [comment]
And applied the two suggestions which were still available.

**Review by AlexWaygood — APPROVED** [review]
LGTM now, thank you!

**Comment by hugovk** [comment]
Thank you very much for the reviews!

**Pull request merged**